src/main.rs

```diff
@@ -17,63 +18,109 @@ fn main() {
+                        config::Als::Iio { path, thresholds } => Box::new(
+                            als::iio::Als::new(&path, thresholds)
+                                .expect("Unable to initialize ALS IIO sensor"),
+                        ),
+                        config::Als::Time { ref hour_to_lux } => {
+                            Box::new(als::time::Als::new(hour_to_lux))
+                        }
+                        config::Als::Webcam { video, thresholds } => Box::new({
+                            let (webcam_tx, webcam_rx) = mpsc::channel();
+                            std::thread::spawn(move || {
+                                als::webcam::Webcam::new(webcam_tx, video).run();
+                            });
+
+                            als::webcam::Als::new(webcam_rx, thresholds)
+                        }),
+                        config::Als::None => Box::new(als::none::Als::default()),
+                    };
+
+                    let frame_processor: Box<dyn frame::processor::Processor> =
+                        match config.frame.processor {
+                            config::Processor::Vulkan => Box::new(
+                                frame::processor::vulkan::Processor::new()
+                                    .expect("Unable to initialize Vulkan"),
+                            ),
+                        };
+
+                    let frame_capturer: Box<dyn frame::capturer::Capturer> =
+                        match config.frame.capturer {
+                            config::Capturer::Wlroots => {
+                                Box::new(frame::capturer::wlroots::Capturer::new(frame_processor))
+                            }
+                            config::Capturer::None => {
+                                Box::new(frame::capturer::none::Capturer::default())
+                            }
+                        };
+
+                    match capturer_config {
+                        config::Output::Backlight(cfg) => {
+                            let controller = predictor::Controller::new(
+                                prediction_tx,
+                                user_rx,
+                                als,
+                                true,
+                                cfg.model.clone(),
+                            );
+                            frame_capturer.run(cfg.model, controller)
+                        }
+                        config::Output::DdcUtil(cfg) => {
+                            let controller = predictor::Controller::new(
+                                prediction_tx,
+                                user_rx,
+                                als,
+                                true,
+                                cfg.model.clone(),
+                            );
+                            frame_capturer.run(cfg.model, controller)
+                        }
+                    };
+                }),
```

Comment:
ideally this piece of code would look like this, for it to happen you'd also have to change a few places from accepting `String` to accepting `&str` (it's a good practice exactly for this purpose, to be able to send the same string to multiple places without lot of memory copy)

```rust
let model = match capturer_config {
    config::Output::Backlight(cfg) => &cfg.model,
    config::Output::DdcUtil(cfg) => &cfg.model,
};

let controller =
    predictor::Controller::new(prediction_tx, user_rx, als, true, model);
frame_capturer.run(model, controller)
```

Reply:
not exactly like this but fixed :)